feat(registry): new CRD option

mloiseleur opened the pull request:
**Description**

This is a follow up of https://github.com/kubernetes-sigs/external-dns/pull/4598, updated with `DNSRecord` name and rebased on current master code. 

The use of an other Kubeconfig will be for an other PR. Same for chart integration, with required new RBAC.

Fixes #4575 

**Checklist**

- [x] Unit tests updated
- [ ] End user documentation updated

ivankatliarchuk commented:
Any chance we keep all API versions in folders `apis/v1alpha`?

Example https://github.com/gofogo/k8s-sigs-external-dns-fork/pull/15/files. This PR is in my repo still, it is working, but I have too many PR currently open ;-)

And API GW https://github.com/kubernetes-sigs/gateway-api/tree/main/apis.

The convention is for Custom Resources to keep them under structure `apis/v1` `apis/v1alpha` and etc. This will pay off with versioning in future

mloiseleur commented:
> Any chance we keep all API versions in folders `apis/v1alpha`?

I was thinking of doing something like that. That make sense to me :+1: .

stevehipwell commented on charts/external-dns/crds/dnsrecord.yaml:
```diff
@@ -0,0 +1,97 @@
+---
+apiVersion: apiextensions.k8s.io/v1
```
```suggestion
apiVersion: apiextensions.k8s.io/v1
```
Nit, we only have document separators where required.

stevehipwell commented on charts/external-dns/crds/v1alpha1.yaml:
```diff
@@ -101,3 +101,100 @@ spec:
     storage: true
     subresources:
       status: {}
+---
```
The CRDs for the Helm chart should be in separate files and be formatted as described in #5479. The following command would be my suggestion to handle multiple CRDs being generated.

```shell
${CONTROLLER_GEN} object crd:crdVersions=v1 paths="./apis/..." output:crd:stdout | yamlfmt - | yq eval '.metadata.annotations |= with_entries(select(.key | test("kubernetes\.io")))' --split-exp '"./charts/external-dns/crds/" + .metadata.name + ".yaml"'
```

ivankatliarchuk commented:
It probably makes sense to slice this PR into smaller parts. There's no major issue if code is added but not immediately functional. For instance, we could safely introduce code from apid/v1alpha1 without yet generating any YAML files. This approach might significantly speed up the implementation of this extremely cool feature!

ivankatliarchuk commented on controller/execute.go:
```diff
@@ -382,9 +382,13 @@ func configureLogger(cfg *externaldns.Config) {
+	case "aws-sd":
+		reg, err = registry.NewAWSSDRegistry(p, cfg.TXTOwnerID)
+	case "crd":
+		reg, err = registry.NewCRDRegistry(p, cfg.KubeConfig, cfg.APIServerURL, cfg.CRDSourceAPIVersion, cfg.Namespace, cfg.TXTOwnerID, cfg.RequestTimeout, cfg.TXTCacheInterval)
```
could we have something like 
```
registry.NewCRDRegistry(p, WithKubeConfig(), WithApiServerUrl(), With...)
```

Easier to support and extend, alternative `registry.NewCRDRegistry(p, cfg)`, and manage requried fields within the registry. Not a great, as there is coupling but still

ivankatliarchuk replied:
Something like
```
type RegistryOptions struct {
    Provider            provider.Provider
    KubeConfig          string
    APIServerURL        string
    CRDSourceAPIVersion string
    Namespace           string
    TXTOwnerID          string
    RequestTimeout      time.Duration
    TXTCacheInterval    time.Duration
    // Add new fields here as needed
}

func NewRegistry(opts RegistryOptions) (Registry, error) {
    // use opts.Provider, opts.KubeConfig, etc.
}
```

ivankatliarchuk replied:
opton example
```go
type RegistryOption func(*Registry)

func WithKubeConfig(cfg string) RegistryOption { ... }

func NewRegistry(provider provider.Provider, opts ...RegistryOption) (Registry, error) {
    reg := &CRDRegistry{Provider: provider}
    for _, opt := range opts {
        opt(reg)
    }
    // ...
}
```

ivankatliarchuk replied:
I could add this in another PR, so you just rebase or implement as part of this change and I make changes in other registries

mloiseleur replied:
Yes, that's a good idea, but probably better in a dedicated PR.

mloiseleur commented:
/remove-lifecycle stale

mloiseleur commented:
/remove-lifecycle rotten

mloiseleur committed:
feat(registry): new CRD option

mloiseleur committed:
code & test self review

mloiseleur commented:
@ivankatliarchuk I managed to get a full add / update / delete cycle working locally. This PR can now be reviewed. Please let me know if you still think we should split it. 
Now that I'm more confident with the data model of the CRD, I'm open to slice this PR into multiple small PRs for easier review.

ivankatliarchuk commented:
Yeaah. This is quite cool feature

ivankatliarchuk commented:
The only what we should do different - Not to use Rest but some other appoarch like https://github.com/kubernetes-sigs/external-dns/blob/master/source/crd.go#L53-L54

Othwerise, every each read/write will be using API credits

mloiseleur committed:
align crd registry with others in the factory

mloiseleur committed:
change from REST API to client/cache pattern

mloiseleur commented:
@ivankatliarchuk Thanks, that makes sense. I just pushed this change. Is this what you had in mind ? Any other comments ?

ivankatliarchuk commented:
Is it actually working end-2-end?

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+	dnsname := strings.ReplaceAll(record.DNSName, ".", "-")
+	dnsrecord := &apiv1alpha1.DNSRecord{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      strings.ToLower(fmt.Sprintf("%s-%s", dnsname, record.RecordType)),
```
The generated CRD name is {dnsname-dashes}-{recordType} but omits SetIdentifier. With Route53 weighted/latency policies, multiple records share the same name+type but different SetIdentifiers - these would collide

Example - Creates "sub-example-io-a" for both setIdentifier="eu" and setIdentifier="us".

ivankatliarchuk replied:
`{dnsname-with-dashes}-{recordType}` or `{recordType}-{dnsname-with-dashes}`?

ivankatliarchuk replied:
Potential length overflow for long DNS names. We have 256 characters, just an observation

mloiseleur replied:
I addressed it with 09d2006f21823985fd8f60537edbe6016bde6898 : <dashed-dns-prefix>-<short hash of the full identity> + using columns feature of CRD to improve UX with kubectl

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+// createDNSRecord create a new DNSRecord with k8s API
+func (cr *CRDRegistry) createDNSRecord(ctx context.Context, record *endpoint.Endpoint) error {
+	// name has to follow rfc 1123
+	dnsname := strings.ReplaceAll(record.DNSName, ".", "-")
```
This is a bit I'm not sure

The name is built by replacing `.` with `-` in the DNS name, then appending the record type. Two different DNS names can produce the same result

  - sub.example.io → replace dots → sub-example-io → final name: sub-example-io-a
  - sub-example.io → replace dots → sub-example-io → final name: sub-example-io-a ← same

 Both are valid DNS names but they generate the identical Kubernetes object name. When the second one is created, k8sErrors.IsAlreadyExists is true and the error is silently swallowed - so the second record is never stored, and there's no indication anything went wrong.

mloiseleur replied:
I addressed it with 09d2006f21823985fd8f60537edbe6016bde6898 : - + using columns feature of CRD to improve UX with kubectl

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+		}
+	}
+
+	err := cr.provider.ApplyChanges(ctx, filteredChanges)
```
CRD writes before provider apply.  The loop creates/deletes/updates CRD objects before calling cr.provider.ApplyChanges. If the provider fails mid-way, the CRDs are already in an inconsistent state with no rollback path.

Need to understand how the CRD represented in kubernetes. 

currently

```
type DNSRecordStatus struct {
	// The generation observed by the external-dns controller.
	// +optional
	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
}
```

It has lack of anything like 
```
	// ReasonAccepted indicates the endpoint has been accepted
	ReasonAccepted ConditionReason = "Accepted"

	// ReasonProgrammed indicates successful programming to DNS provider
	ReasonProgrammed ConditionReason = "Programmed"
    ^ records successfuly applied
   
    // ReasonFailed indicates one or more records failed to provision
	ReasonFailed ConditionReason = "Failed"
   ^ records failed to be applied
```

mloiseleur replied:
To me, it was the next step, but your comment motivated me to implement it right away.
There is now a Status on this CRD which can be Accepted, Programmed or Failed, see 3c225f403799c8d1134f40548845355e6e9d2fb2

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+limitations under the License.
+*/
+
+package registry
```
```suggestion
package crd
```

ivankatliarchuk replied:
Package name conflicts with parent.

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on registry/crd/crd_test.go:
```diff
@@ -0,0 +1,324 @@
+limitations under the License.
+*/
+
+package registry
```
```suggestion
package crd
```

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+	ownerID   string // refers to the owner id of the current instance
+}
+
+func New(cfg *externaldns.Config, p provider.Provider) (registry.Registry, error) {
```
TXTOwnerID reused for CRD registry. Not something we should immediately change, more a cosmetic. There should be a shared OwnerID field most likely, as we reusing it.

mloiseleur replied:
Good point :+1: . I've updated flags doc to make this more clear (4c53b4d28a982cef4705b38f6ced85ab74d201b7)

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,345 @@
+	"sigs.k8s.io/external-dns/plan"
+	"sigs.k8s.io/external-dns/provider"
+	"sigs.k8s.io/external-dns/registry"
+	"sigs.k8s.io/external-dns/source"
```
source package dependency from registry. Registry code depending on the source package creates a bidirectional dependency: source → registry already exists, adding registry/crd → source creates a cycle risk.

ivankatliarchuk replied:
In external-dns the intended dependency flow is:

```
  controller → registry → provider
  controller → source  → provider
```

 source and registry are siblings — neither should depend on the other.

What this PR does

```
  registry/crd/crd.go imports source solely to get SingletonClientGenerator:

  import "sigs.k8s.io/external-dns/source"

  clientGenerator := &source.SingletonClientGenerator{
      KubeConfig:     kubeConfig,
      APIServerURL:   apiServerURL,
      RequestTimeout: apiServerTimeout,
  }
  restConfig, err := clientGenerator.RESTConfig()
```

Not sure if it worth to overcomplicate this PR, but looks like we need to move RESTConfig() to a shared package `pkg/client`  that both source and registry can import without
  either depending on the other.

mloiseleur replied:
Yes, I'll probably do it in another, follow up PR.

ivankatliarchuk replied:
k

ivankatliarchuk commented on apis/v1alpha1/dnsrecord.go:
```diff
@@ -0,0 +1,92 @@
+// DNSRecord is used to get all records managed by external-dns.
+// It can be used as a registry with the status subresource.
+// +k8s:openapi-gen=true
+// +groupName=externaldns.k8s.io
```
+groupName is a package-level annotation that belongs in doc.go, not on the type definition.

mloiseleur replied:
fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on apis/v1alpha1/dnsrecord.go:
```diff
@@ -0,0 +1,92 @@
+		spec.SetIdentifier == e.SetIdentifier
+}
+
+func (dr *DNSRecord) EndpointLabels() endpoint.Labels {
```
The EndpointLabels() method is defined on DNSRecord but never called anywhere in the codebase. What is for?

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on apis/v1alpha1/dnsrecord.go:
```diff
@@ -0,0 +1,92 @@
+	Items           []DNSRecord `json:"items"`
+}
+
+func (dr *DNSRecord) IsEndpoint(e *endpoint.Endpoint) bool {
```
Same. Looks like IsEndpoint is dead code right now, not sure, maybe forget to use it?

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on registry/crd/crd_test.go:
```diff
@@ -0,0 +1,324 @@
+				seedEndpoints = append(seedEndpoints, tc.ep)
+			case "Update":
+				changes.UpdateNew = []*endpoint.Endpoint{tc.ep}
+				changes.UpdateOld = []*endpoint.Endpoint{tc.ep}
```
Is this should be a separate "old" endpoint? UpdateOld should contain the pre-change endpoint to properly exercise the lookup path

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on apis/v1alpha1/dnsrecord.go:
```diff
@@ -0,0 +1,92 @@
+// DNSRecordSpec defines the desired state of DNSEndpoint
+// +kubebuilder:object:generate=true
+type DNSRecordSpec struct {
+	Endpoint endpoint.Endpoint `json:"endpoints,omitempty"`
```
`endpoints` > plural tag on singular field

ivankatliarchuk replied:
Maybe should  be 
```
type DNSRecordSpec struct {
   Endpoints ...
}
```

I don't know, just guessing

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk commented on apis/v1alpha1/dnsrecord.go:
```diff
@@ -0,0 +1,92 @@
+	RecordResourceLabel      string = "externaldns.k8s.io/resource"
+)
+
+// DNSRecordSpec defines the desired state of DNSEndpoint
```
```suggestion
// DNSRecordSpec defines the desired state of Endpoint
```

or `DNSRecord`

As DNSEndpoint is a separate, other CRD that already exists in the codebase, and they not related.

mloiseleur replied:
Fixed with ad2bd93522fa70cc61e4fe97a87e33323e4c481c

ivankatliarchuk reviewed (changes requested):
Have a look what worth changing, and what you planning to address in follow up

mloiseleur committed:
refactor(registry/crd): address review feedback

- rename package registry -> crd to match its directory and drop the
  now-redundant import alias in the factory
- remove dead IsEndpoint() and EndpointLabels() methods
- fix DNSRecordSpec doc comment (DNSEndpoint -> DNSRecord) and drop the
  redundant +groupName marker already declared at package level
- rename the singular field's JSON tag endpoints -> endpoint and
  regenerate the CRD manifests
- exercise a distinct UpdateOld endpoint in the ApplyChanges update test

mloiseleur committed:
fix(registry/crd): prevent DNSRecord name collisions

The object name was built as <dashed-dns-name>-<type>, which collapsed
distinct endpoints onto the same name: different set identifiers
(Route53 weighted/latency), sub.example.io vs sub-example.io, and names
truncated to the length limit. Colliding creates were silently swallowed
via IsAlreadyExists, so records were never stored.

Build a deterministic, RFC 1123 compliant name from a readable dashed DNS
prefix plus a short hash of the full identity (DNS name, record type, set
identifier). The dashed prefix keeps records discoverable with
`kubectl get dnsrecords`; the hash guarantees uniqueness.

Record identity now lives in the object name instead of labels, so lookups
Get-by-name rather than match labels. This also fixes DNS names longer than
63 characters, which previously overflowed the label-value limit and broke
create/lookup/delete. Only the owner label remains.

Add printer columns (DNS name, type, set identifier, targets) so the full
domain stays visible regardless of the object name.

mloiseleur committed:
docs(registry/crd): note CRD registry uses txt-owner-id

The --txt-owner-id flag also identifies the instance for the CRD registry,
but its help text only mentioned TXT and DynamoDB. Add CRD so users know
which flag sets the owner id for it.

mloiseleur committed:
feat(registry/crd): record Accepted, Programmed and Failed status

ExternalDNS persists each managed endpoint as a DNSRecord. To make the
outcome of a reconcile visible on the object — and to keep the CRD store a
trustworthy view of what was applied — record the full lifecycle as status
conditions.

Intent is recorded first: the DNSRecord objects for created and updated
endpoints are written and marked Accepted before the provider is called.
Their Programmed condition is then set from the provider outcome: True on
success, or False with reason Failed when the provider rejects the change.

The provider returns a single batch error and cannot attribute it to
individual records, so on failure every record in the batch is marked
Failed. Provider applies are idempotent, so any record that was in fact
applied is corrected to Programmed on the next reconcile.

Records() — the plan's view of current state — now returns only records
whose Programmed condition is True. This is what keeps writing the DNSRecord
before the provider apply safe: a record left Accepted by a provider failure
is not mistaken for an existing record, so the plan re-applies it on the next
reconcile, never leaving a DNSRecord claiming a record the provider rejected
(which the plan would treat as present and never retry).

Status updates are best-effort: a failure to write status is logged but does
not fail reconciliation, since the DNS record itself is already applied (or
its failure already surfaced through the apply error). The Accepted and
Programmed conditions are surfaced as kubectl print columns.

mloiseleur committed:
docs(registry/crd): document the CRD registry

Add a usage page for the CRD registry: installing the DNSRecord CRD,
required RBAC, configuration flags, the Programmed status condition, and
its current in-cluster-only limitation. Link it from the registry index
and the docs navigation.

mloiseleur commented:
@ivankatliarchuk Thanks for this complete review. I've tried to address most of them + added a doc page on this registry.

> Is it actually working end-2-end?

I tested locally the first reviewable version with coredns provider on k3d (thanks to your great tutorial).

I'll probably do another full local test run this weekend, to see all those changes in action!

mloiseleur committed:
self-review

mloiseleur commented:
@ivankatliarchuk I re-tested locally with coredns following this [tutorial](https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/coredns-etcd.md). Everything went fine:

<details>
<summary>Records creation</summary>
```sh
DEBU[0000] Endpoints generated from service: default/a-g1-record: [a.example.org 0 IN A  172.18.0.2 [{coredns/group g1}]] 
DEBU[0000] Endpoints generated from service: default/aa-g1-record: [aa.example.org 0 IN AAAA  2001:db8::1 [{coredns/group g1}]] 
DEBU[0000] No endpoints could be generated from 'service/default/coredns' 
DEBU[0000] No endpoints could be generated from 'service/default/etcd' 
DEBU[0000] No endpoints could be generated from 'service/default/etcd-nodeport-external' 
DEBU[0000] No endpoints could be generated from 'service/default/kubernetes' 
INFO[0000] Generating new prefix: (3dc2be40)            
INFO[0000] Delete key /skydns/org/example/aa/default    
INFO[0000] Add/set key /skydns/org/example/aa/3dc2be40 to Host=2001:db8::1, Text=, TTL=0 
INFO[0000] Generating new prefix: (25d8864c)            
INFO[0000] Delete key /skydns/org/example/a/default     
INFO[0000] Add/set key /skydns/org/example/a/25d8864c to Host=172.18.0.2, Text=, TTL=0 
DEBU[0000] Getting service (&{172.18.0.2 0 10 0  false 0 1 g1 /skydns/org/example/a/25d8864c }) with service host (172.18.0.2) 
DEBU[0000] Creating new ep (a.example.org 0 IN A  172.18.0.2 [{coredns/group g1}]) with new service host (172.18.0.2) 
DEBU[0000] Getting service (&{2001:db8::1 0 10 0  false 0 1 g1 /skydns/org/example/aa/3dc2be40 }) with service host (2001:db8::1) 
DEBU[0000] Creating new ep (aa.example.org 0 IN A  2001:db8::1 [{coredns/group g1}]) with new service host (2001:db8::1) 
DEBU[0000] update DNSRecord with modified labels from provider 
DEBU[0000] no provider record matched DNSRecord aa-example-org-c71d29b5; skipping label adjust 
DEBU[0030] dedupSource: collecting endpoints and removing duplicates 
DEBU[0030] multiSource: collecting endpoints from 1 child sources and removing duplicates 
DEBU[0030] No endpoints could be generated from 'service/default/etcd' 
DEBU[0030] No endpoints could be generated from 'service/default/etcd-nodeport-external' 
DEBU[0030] No endpoints could be generated from 'service/default/kubernetes' 
DEBU[0030] No endpoints could be generated from 'service/kube-system/kube-dns' 
DEBU[0030] Endpoints generated from service: default/a-g1-record: [a.example.org 0 IN A  172.18.0.2 [{coredns/group g1}]] 
DEBU[0030] Endpoints generated from service: default/aa-g1-record: [aa.example.org 0 IN AAAA  2001:db8::1 [{coredns/group g1}]] 
DEBU[0030] No endpoints could be generated from 'service/default/coredns' 
INFO[0030] All records are already up to date           
```

```sh
$ kubectl exec -it etcd-0 -- etcdctl get /skydns/org/example --prefix --keys-only
/skydns/org/example/a/25d8864c

/skydns/org/example/aa/3dc2be40

$ k get dnsrecord
NAME                      DNS NAME         TYPE   SET ID   TARGETS           STATUS
a-example-org-19d9a713    a.example.org    A               ["172.18.0.2"]    Programmed
aa-example-org-c71d29b5   aa.example.org   AAAA            ["2001:db8::1"]   Programmed

$ k describe dnsrecord a-example-org-19d9a713 
Name:         a-example-org-19d9a713
Namespace:    default
Labels:       externaldns.k8s.io/owner=default
Annotations:  <none>
API Version:  externaldns.k8s.io/v1alpha1
Kind:         DNSRecord
Metadata:
  Creation Timestamp:  2026-06-06T14:59:52Z
  Generation:          2
  Resource Version:    1204
  UID:                 08c54bb5-0c2a-42d7-a7bb-7a70d737b6f1
Spec:
  Endpoint:
    Dns Name:  a.example.org
    Labels:
      172.18.0.2:     25d8864c
      Original Text:  
      Owner:          default
      Prefix:         25d8864c
      Resource:       service/default/a-g1-record
    Provider Specific:
      Name:       coredns/group
      Value:      g1
    Record Type:  A
    Targets:
      172.18.0.2
Status:
  Conditions:
    Last Transition Time:  2026-06-06T14:59:52Z
    Message:               Endpoint applied to the DNS provider
    Reason:                Programmed
    Status:                True
    Type:                  Ready
Events:                    <none>
```
</details>

<details>

<summary>Records update</summary>

```sh
$ kubectl patch svc a-g1-record --type=merge  -p '{"status":{"loadBalancer":{"ingress":[{"ip":"172.18.0.4"}]}}}'   --subresource=status
service/a-g1-record patched
```

```sh
DEBU[0092] dedupSource: collecting endpoints and removing duplicates 
DEBU[0092] multiSource: collecting endpoints from 1 child sources and removing duplicates 
DEBU[0092] Endpoints generated from service: default/aa-g1-record: [aa.example.org 0 IN AAAA  2001:db8::1 [{coredns/group g1}]] 
DEBU[0092] No endpoints could be generated from 'service/default/coredns' 
DEBU[0092] No endpoints could be generated from 'service/default/etcd' 
DEBU[0092] No endpoints could be generated from 'service/default/etcd-nodeport-external' 
DEBU[0092] No endpoints could be generated from 'service/default/kubernetes' 
DEBU[0092] No endpoints could be generated from 'service/kube-system/kube-dns' 
DEBU[0092] Endpoints generated from service: default/a-g1-record: [a.example.org 0 IN A  172.18.0.4 [{coredns/group g1}]] 
DEBU[0092] Updating labels (map[owner:default resource:service/default/a-g1-record]) with old labels (map[172.18.0.2:25d8864c originalText: owner:default prefix:25d8864c resource:service/default/a-g1-record]) 
INFO[0092] Generating new prefix: (130556c9)            
INFO[0092] Delete key /skydns/org/example/a/25d8864c    
INFO[0092] Delete key /skydns/org/example/a/default     
INFO[0092] Add/set key /skydns/org/example/a/130556c9 to Host=172.18.0.4, Text=, TTL=0 
DEBU[0092] Getting service (&{172.18.0.4 0 10 0  false 0 1 g1 /skydns/org/example/a/130556c9 }) with service host (172.18.0.4) 
DEBU[0092] Creating new ep (a.example.org 0 IN A  172.18.0.4 [{coredns/group g1}]) with new service host (172.18.0.4) 
DEBU[0092] Getting service (&{2001:db8::1 0 10 0  false 0 1 g1 /skydns/org/example/aa/3dc2be40 }) with service host (2001:db8::1) 
DEBU[0092] Creating new ep (aa.example.org 0 IN A  2001:db8::1 [{coredns/group g1}]) with new service host (2001:db8::1) 
DEBU[0092] update DNSRecord with modified labels from provider 
```

```sh
$ k get dnsrecord
NAME                      DNS NAME         TYPE   SET ID   TARGETS           STATUS
a-example-org-19d9a713    a.example.org    A               ["172.18.0.4"]    Programmed
aa-example-org-c71d29b5   aa.example.org   AAAA            ["2001:db8::1"]   Programmed
```

</details>

<details>

<summary>Records delete</summary>

```sh
$ kubectl delete svc -l svc=test-svc
service "a-g1-record" deleted
service "aa-g1-record" deleted
```

```sh
DEBU[0122] dedupSource: collecting endpoints and removing duplicates 
DEBU[0122] multiSource: collecting endpoints from 1 child sources and removing duplicates 
DEBU[0122] No endpoints could be generated from 'service/default/coredns' 
DEBU[0122] No endpoints could be generated from 'service/default/etcd' 
DEBU[0122] No endpoints could be generated from 'service/default/etcd-nodeport-external' 
DEBU[0122] No endpoints could be generated from 'service/default/kubernetes' 
DEBU[0122] No endpoints could be generated from 'service/kube-system/kube-dns' 
INFO[0122] Delete key /skydns/org/example/a/130556c9    
INFO[0122] Delete key /skydns/org/example/aa            
DEBU[0153] dedupSource: collecting endpoints and removing duplicates 
DEBU[0153] multiSource: collecting endpoints from 1 child sources and removing duplicates 
DEBU[0153] No endpoints could be generated from 'service/default/etcd' 
DEBU[0153] No endpoints could be generated from 'service/default/etcd-nodeport-external' 
DEBU[0153] No endpoints could be generated from 'service/default/kubernetes' 
DEBU[0153] No endpoints could be generated from 'service/kube-system/kube-dns' 
DEBU[0153] No endpoints could be generated from 'service/default/coredns' 
INFO[0153] All records are already up to date           
```

```sh
$ k get dnsrecord
No resources found in default namespace.
```

```sh
kubectl exec -it etcd-0 -- etcdctl get /skydns/org/example --prefix --keys-only
```

</details>

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,483 @@
+	return nil
+}
+
+func (cr *CRDRegistry) updateDNSRecordWithEndpointLabels(ctx context.Context, dnsrecord *apiv1alpha1.DNSRecord, record *endpoint.Endpoint) error {
```
This kind of methods usually lives in apis https://github.com/gofogo/k8s-sigs-external-dns-fork/pull/58/changes#diff-a9f7e5ea46601edff40422542f00754183019d066d2b01052b2d4df1312567cc

mloiseleur replied:
Yep, fixed with 8d5407b4e44337ed5b4fc3579e660941b1a7c39d

ivankatliarchuk commented on pkg/apis/externaldns/types.go:
```diff
@@ -180,8 +180,8 @@ type Config struct {
 	ExoscaleZoneCacheDuration                     time.Duration
-	CRDSourceAPIVersion                           string
-	CRDSourceKind                                 string
+	CRDAPIVersion                                 string
```
What this flags are for? They are harmful. I know they have advantages, but we should consider be following well known kubernetes version support `v1alpha -> v1beta -> v1. or `v1alpha -> v1`. Another question, let's say now I could do 

```
--crd-source-apiversion externaldns.my-exceptions.io/v1
--crd-source-kind My3rdPartyKind
```

^ How this is going to work? It will create noice in issues, while we only support at the moment specific apiversion and kind.

mloiseleur replied:
I kind agree with you. When I discovered those 2 flags, I was a bit perplex about them. I was focused on the registry, so I just did a rename.
FTM, I reverted the rename in 8d5407b4e44337ed5b4fc3579e660941b1a7c39d, so this PR does not touch them. We can think and maybe remove those flags in a dedicated follow up PR.

ivankatliarchuk commented on registry/crd/crd.go:
```diff
@@ -0,0 +1,483 @@
+	// Stamp the owner label before applying so the provider sees owner-labeled
+	// endpoints, matching the TXT registry behavior.
+	for _, r := range filteredChanges.Create {
+		if r.Labels == nil {
```
method https://github.com/kubernetes-sigs/external-dns/blob/50083e5807c15858892d95bc973703bb99bbff51/endpoint/endpoint.go#L453

```
r.WithLabel(endpoint.OwnerLabelKey, cr.ownerId)
```

mloiseleur replied:
thanks, implemented with 8d5407b4e44337ed5b4fc3579e660941b1a7c39d

mloiseleur committed:
address ivan review comments

mloiseleur commented:
@ivankatliarchuk I've addressed your review comments. 

I've 2 follow-up PRs in sight:

1. https://github.com/kubernetes-sigs/external-dns/pull/5372#discussion_r3350866756
2. https://github.com/kubernetes-sigs/external-dns/pull/5372#discussion_r3369859504

ivankatliarchuk commented:
/lgtm

mloiseleur commented:
/approve

mloiseleur commented:
/ok-to-test

stevehipwell commented:
@mloiseleur the chart shouldn't be updated until there is a released External DNS version that supports the changes, and nor should it be updated without a CHANGELOG entry. Unless there is something I'm missing it looks like you will need to revert the chart changes?

mloiseleur commented:
@stevehipwell Maybe you wanted to comment this PR ? https://github.com/kubernetes-sigs/external-dns/pull/6513 

:thinking: or you wanted to say that this PR adds the CRD to the chart, without changelog entry?

stevehipwell commented:
@mloiseleur I don't have the full context of the CRD in question, but based on the PR where it was added it looks like it's not part of the currently released version of External DNS? We have a policy of not accepting chart changes until the feature they're supporting has been released, and we always require a CHANGELOG entry. I did comment to this effect on one of your previous PRs where you were working on the CRD.

mloiseleur commented:
Here it is: https://github.com/kubernetes-sigs/external-dns/pull/6516

stevehipwell commented:
> Here it is: https://github.com/kubernetes-sigs/external-dns/pull/6516

Thanks @mloiseleur. I'm happy to prioritise reviewing/approving the PR to add it when it's ready. I just don't want to be in the position where we accidentally release a chart patch with changes for the next version.
